File: Scala/main/chap15/Section_6.scala
package chap15

object Section_6:
   trait Publication:
      def title: String
      def pageCount: Int

   case class Book(title: String, author: String, pageCount: Int)  extends Publication
   case class Magazine(title: String, number: Int, pageCount: Int) extends Publication

   def printTitle(publication: Publication): Unit = println(publication.title)

   // rejected by the compiler
   // case class Book(author: String, pageCount: Int) extends Publication

   object Variant1:
      case class Book(author: String, pageCount: Int) // OK
      val book: Book = mocks.Chap15.noTitleBook
      // printTitle(book) // rejected by the compiler

   object Variant2:
      case class Book(title: String, author: String, pageCount: Int)
      val book: Book = mocks.Chap15.notPubBook
      // printTitle(book) // rejected by the compiler

File: Python/chap15/Section_6.py
#!/usr/bin/env python3

from Section_1 import Book, print_title


book = Book(title="Le Comte de Monte-Cristo", author="Alexandre Dumas", pagecount=1476)

print_title(book) # prints "Le Comte de Monte-Cristo"

class Magazine:
    def __init__(self, title, number, pagecount):
        self.title = title
        self.number = number
        self.pagecount = pagecount


magazine = Magazine(title="Life", number=123, pagecount=45)
print_title(magazine) # prints "Life"

class Noble:
    def __init__(self, name, title):
        self.name = name
        self.title = title


person = Noble(name="Edmond Dantès", title="Comte de Monte-Cristo")
print_title(person) # prints "Comte de Monte-Cristo"
